Mathlib/Algebra/Group/Pi/Sum.lean

```diff
@@ -0,0 +1,110 @@
+/-
+Copyright (c) 2024 Amelia Livingston. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Amelia Livingston
+-/
+
+import Mathlib.Algebra.Group.Defs
+
+/-!
+# Monoid, group etc structures on families indexed by sum types
+
+Given 2 families of e.g. groups `αᵢ, βᵢ` indexed by `ια, ιβ` respectively, the components
+of the resulting family of types indexed by `ια ⊕ ιβ` each have a group structure.
+
+# Implementation notes
+
+A little care is required to avoid non-defeq diamonds - e.g. `instElimMulOneClass` can't just be
+defined with a match expression, since the resulting `One` and `Mul` instances won't be defeq to
+`instElimOne` and `instElimMul`.
+
+# Todo
+
+Missing some of the algebraic hierarchy compared to e.g. `Mathlib.Algebra.Group.Prod`.
+
+-/
+universe v u
+
+namespace Sum
+
+variable {ια ιβ : Type*} {α : (i : ια) → Type u} {β : (i : ιβ) → Type u}
+
+@[to_additive] instance instElimOne [∀ (i : ια), One (α i)] [∀ i : ιβ, One (β i)]
+    (i : ια ⊕ ιβ) : One (Sum.elim α β i) :=
+  match i with
+  | Sum.inl i => inferInstanceAs (One (α i))
+  | Sum.inr i => inferInstanceAs (One (β i))
```

Comment:
I think it would be best to define this as
```suggestion
  { one := match i with
    | Sum.inl i => (1 : α i)
    | Sum.inr i => (1 : β i) }
```
to allow things to be unfolded slightly further.